internal/tuning.go

```diff
@@ -0,0 +1,395 @@
+// The MIT License
+//
+// Copyright (c) 2022 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"golang.org/x/sync/semaphore"
+
+	"go.temporal.io/sdk/internal/common/metrics"
+	"go.temporal.io/sdk/log"
+)
+
+// WorkerTuner allows for the dynamic customization of some aspects of worker behavior.
+//
+// WARNING: Custom implementations of SlotSupplier are currently experimental.
+type WorkerTuner interface {
+	GetWorkflowTaskSlotSupplier() SlotSupplier
+	GetActivityTaskSlotSupplier() SlotSupplier
+	GetLocalActivitySlotSupplier() SlotSupplier
+	GetNexusSlotSupplier() SlotSupplier
+}
+
+// SlotPermit is a permit to use a slot.
+//
+// WARNING: Custom implementations of SlotSupplier are currently experimental.
+type SlotPermit struct {
+	//lint:ignore U1000 pointless to guarantee pointers to SlotPermits are unique
+	_uniqueInt int
+}
+
+// SlotReserveContext contains information that SlotSupplier instances can use during
+// reservation calls. It embeds a standard Context.
+type SlotReserveContext interface {
+	context.Context
```

Comment:
We should not create a custom context type. That is a accepted Go idiom https://go.dev/wiki/CodeReviewComments#contexts

Reply:
I was just going with Chad's suggestion here: https://github.com/temporalio/sdk-go/pull/1546#discussion_r1693589674

I don't care much either way. I would be happy to do my original thing where the std context is separate.

Reply:
Yeah I would not embedded the context, I would keep them separate. I am not sure about this `Slot*Context ` , it has an established meaning in Go, maybe `Slot*Request`?

Reply:
Coming back here, the separation of the two contexts and not calling the latter a context to avoid confusion makes sense to me